File: src/main/kotlin/io/titandata/titan/providers/local/Checkout.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.providers.local

import io.titandata.client.apis.CommitsApi
import io.titandata.client.apis.RepositoriesApi
import io.titandata.titan.clients.Docker
import io.titandata.titan.utils.CommandExecutor

class Checkout(
    private val commandExecutor: CommandExecutor = CommandExecutor(),
    private val docker: Docker = Docker(commandExecutor),
    private val commitsApi: CommitsApi = CommitsApi(),
    private val repositoriesApi: RepositoriesApi = RepositoriesApi()
) {
    fun checkout(container: String, guid: String?, tags: List<String>) {
        val sourceCommit = if (guid == null) {
            if (tags.isNotEmpty()) {
                val commits = commitsApi.listCommits(container, tags)
                if (commits.size == 0) {
                    throw IllegalStateException("no matching commits found")
                }
                commits.first().id
            } else {
                val status = repositoriesApi.getRepositoryStatus(container)
                if (status.sourceCommit == null) {
                    throw IllegalStateException("no commits present, run 'titan commit' first")
                }
                status.sourceCommit!!
            }
        } else {
            if (tags.isNotEmpty()) {
                throw IllegalArgumentException("tags and commit cannot both be specified")
            }
            guid
        }

        println("Stopping container $container")
        docker.stop(container)
        println("Checkout $sourceCommit")
        commitsApi.checkoutCommit(container, sourceCommit)
        println("Starting container $container")
        docker.start(container)
        println("$sourceCommit checked out")
    }
}

File: internal/app/providers/local/Checkout.go
package local

import (
	"fmt"
	"os"
	"strconv"
	"titan/internal/app/clients"
)

func Checkout(repo string, guid string, tags[]string, port int, context string) {
	cfg.BasePath = "http://localhost:" + strconv.Itoa(port)
	docker := clients.Docker(context, port)

	var sourceCommit string
	if guid == "" {
		if len(tags) > 0 {
			commits, _, _ := commitsApi.ListCommits(ctx, repo, nil)//TODO pass tags
			if len(commits) == 0 {
				fmt.Println("no matching commits found")
				os.Exit(1)
			}
			sourceCommit = commits[0].Id
		} else {
			status, _, _ := repositoriesApi.GetRepositoryStatus(ctx, repo)
			if status.SourceCommit == "" {
				fmt.Println("no commits present, run 'titan commit' first")
				os.Exit(1)
			}
			sourceCommit = status.SourceCommit
		}
	} else {
		if len(tags) > 0 {
			fmt.Println("tags and commit cannot both be specified")
			os.Exit(1)
		}
		sourceCommit = guid
	}
	fmt.Println("Stopping container " + repo)
	docker.Stop(repo)
	fmt.Println("Checkout " + sourceCommit)
	commitsApi.CheckoutCommit(ctx, repo, sourceCommit)
	fmt.Println("Starting container " + repo)
	docker.Start(repo)
	fmt.Println(sourceCommit + " checked out")
}
